File: internal/runner/runner.go
package runner

import (
	"bytes"
	"context"
	"fmt"
	"io"
	"os"
	"os/exec"
	"strings"
	"time"

	"github.com/keyxmare/DevBootstrap/internal/cli"
)

// Runner executes shell commands with logging and error handling.
type Runner struct {
	CLI     *cli.CLI
	DryRun  bool
	UseSudo bool
}

// New creates a new Runner instance.
func New(c *cli.CLI, dryRun bool) *Runner {
	return &Runner{
		CLI:    c,
		DryRun: dryRun,
	}
}

// Run executes a command and captures its output.
func (r *Runner) Run(args []string, opts ...Option) *Result {
	options := &runOptions{
		timeout: 2 * time.Minute,
	}
	for _, opt := range opts {
		opt(options)
	}

	// Add sudo if needed
	if options.sudo && os.Geteuid() != 0 {
		args = append([]string{"sudo"}, args...)
	}

	cmdStr := strings.Join(args, " ")

	if options.description != "" {
		r.CLI.PrintProgress(options.description)
	}

	if r.DryRun {
		r.CLI.ClearProgress()
		r.CLI.PrintInfo(fmt.Sprintf("[DRY RUN] %s", cmdStr))
		return SuccessResult("")
	}

	ctx, cancel := context.WithTimeout(context.Background(), options.timeout)
	defer cancel()

	cmd := exec.CommandContext(ctx, args[0], args[1:]...)

	// Set environment
	cmd.Env = os.Environ()
	if len(options.env) > 0 {
		for k, v := range options.env {
			cmd.Env = append(cmd.Env, fmt.Sprintf("%s=%s", k, v))
		}
	}

	// Set working directory
	if options.cwd != "" {
		cmd.Dir = options.cwd
	}

	var stdout, stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	err := cmd.Run()

	r.CLI.ClearProgress()

	if ctx.Err() == context.DeadlineExceeded {
		return FailureResult("Command timed out", -1, ctx.Err())
	}

	if err != nil {
		exitCode := -1
		if exitErr, ok := err.(*exec.ExitError); ok {
			exitCode = exitErr.ExitCode()
		}
		return FailureResult(stderr.String(), exitCode, err)
	}

	return SuccessResult(strings.TrimSpace(stdout.String()))
}

// RunInteractive executes a command with real-time output to the terminal.
func (r *Runner) RunInteractive(args []string, opts ...Option) bool {
	options := &runOptions{}
	for _, opt := range opts {
		opt(options)
	}

	// Add sudo if needed
	if options.sudo && os.Geteuid() != 0 {
		args = append([]string{"sudo"}, args...)
	}

	if options.description != "" {
		r.CLI.PrintInfo(options.description)
	}

	if r.DryRun {
		r.CLI.PrintInfo(fmt.Sprintf("[DRY RUN] %s", strings.Join(args, " ")))
		return true
	}

	cmd := exec.Command(args[0], args[1:]...)
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr
	cmd.Stdin = os.Stdin

	// Set environment
	cmd.Env = os.Environ()
	if len(options.env) > 0 {
		for k, v := range options.env {
			cmd.Env = append(cmd.Env, fmt.Sprintf("%s=%s", k, v))
		}
	}

	// Set working directory
	if options.cwd != "" {
		cmd.Dir = options.cwd
	}

	if err := cmd.Run(); err != nil {
		r.CLI.PrintError(fmt.Sprintf("Erreur: %v", err))
		return false
	}

	return true
}

// CommandExists checks if a command exists in PATH.
func (r *Runner) CommandExists(name string) bool {
	_, err := exec.LookPath(name)
	return err == nil
}

// GetCommandPath returns the full path to a command.
func (r *Runner) GetCommandPath(name string) string {
	path, err := exec.LookPath(name)
	if err != nil {
		return ""
	}
	return path
}

// GetCommandVersion returns the version of a command.
func (r *Runner) GetCommandVersion(name string, versionArg string) string {
	if versionArg == "" {
		versionArg = "--version"
	}

	result := r.Run([]string{name, versionArg})
	if result.Success {
		// Return first line
		lines := strings.Split(result.Stdout, "\n")
		if len(lines) > 0 {
			return strings.TrimSpace(lines[0])
		}
	}
	return ""
}

// EnsureDirectory creates a directory if it doesn't exist.
func (r *Runner) EnsureDirectory(path string, mode os.FileMode) error {
	if r.DryRun {
		r.CLI.PrintInfo(fmt.Sprintf("[DRY RUN] mkdir -p %s", path))
		return nil
	}
	return os.MkdirAll(path, mode)
}

// DownloadFile downloads a file from a URL.
func (r *Runner) DownloadFile(url, destination string, opts ...Option) bool {
	options := &runOptions{}
	for _, opt := range opts {
		opt(options)
	}

	if options.description != "" {
		r.CLI.PrintProgress(options.description)
	}

	// Try curl first
	if r.CommandExists("curl") {
		result := r.Run([]string{"curl", "-fsSL", "-o", destination, url}, WithTimeout(10*time.Minute))
		if result.Success {
			r.CLI.ClearProgress()
			return true
		}
	}

	// Fall back to wget
	if r.CommandExists("wget") {
		result := r.Run([]string{"wget", "-q", "-O", destination, url}, WithTimeout(10*time.Minute))
		if result.Success {
			r.CLI.ClearProgress()
			return true
		}
	}

	r.CLI.ClearProgress()
	r.CLI.PrintError("Ni curl ni wget n'est disponible")
	return false
}

// CopyFile copies a file from src to dst.
func (r *Runner) CopyFile(src, dst string) error {
	if r.DryRun {
		r.CLI.PrintInfo(fmt.Sprintf("[DRY RUN] cp %s %s", src, dst))
		return nil
	}

	srcFile, err := os.Open(src)
	if err != nil {
		return err
	}
	defer srcFile.Close()

	dstFile, err := os.Create(dst)
	if err != nil {
		return err
	}
	defer dstFile.Close()

	_, err = io.Copy(dstFile, srcFile)
	return err
}

// RemoveAll removes a file or directory recursively.
func (r *Runner) RemoveAll(path string) error {
	if r.DryRun {
		r.CLI.PrintInfo(fmt.Sprintf("[DRY RUN] rm -rf %s", path))
		return nil
	}
	return os.RemoveAll(path)
}

// runOptions holds options for running commands.
type runOptions struct {
	description string
	sudo        bool
	env         map[string]string
	cwd         string
	timeout     time.Duration
}

// Option is a functional option for Run.
type Option func(*runOptions)

// WithDescription sets the description for the command.
func WithDescription(desc string) Option {
	return func(o *runOptions) {
		o.description = desc
	}
}

// WithSudo runs the command with sudo.
func WithSudo() Option {
	return func(o *runOptions) {
		o.sudo = true
	}
}

// WithEnv sets environment variables for the command.
func WithEnv(env map[string]string) Option {
	return func(o *runOptions) {
		o.env = env
	}
}

// WithCwd sets the working directory for the command.
func WithCwd(cwd string) Option {
	return func(o *runOptions) {
		o.cwd = cwd
	}
}

// WithTimeout sets the timeout for the command.
func WithTimeout(timeout time.Duration) Option {
	return func(o *runOptions) {
		o.timeout = timeout
	}
}

File: docker_installer/utils/runner.py
"""Command execution utilities."""

import subprocess
import os
import shutil
from dataclasses import dataclass
from typing import Optional, Callable
from .cli import CLI


@dataclass
class CommandResult:
    """Result of a command execution."""
    success: bool
    stdout: str
    stderr: str
    return_code: int


class CommandRunner:
    """Execute shell commands with logging and error handling."""

    def __init__(self, cli: CLI, use_sudo: bool = False, dry_run: bool = False):
        """Initialize the command runner."""
        self.cli = cli
        self.use_sudo = use_sudo
        self.dry_run = dry_run

    def run(
        self,
        command: list[str],
        description: Optional[str] = None,
        sudo: Optional[bool] = None,
        capture_output: bool = True,
        env: Optional[dict] = None,
        cwd: Optional[str] = None,
        timeout: Optional[int] = None
    ) -> CommandResult:
        """Run a command and return the result."""
        use_sudo = sudo if sudo is not None else self.use_sudo

        # Prepend sudo if needed
        if use_sudo and os.geteuid() != 0:
            command = ["sudo"] + command

        cmd_str = " ".join(command)

        if description:
            self.cli.print_progress(description)

        if self.dry_run:
            self.cli.clear_progress()
            self.cli.print_info(f"[DRY RUN] {cmd_str}")
            return CommandResult(
                success=True,
                stdout="",
                stderr="",
                return_code=0
            )

        try:
            # Merge environment
            run_env = os.environ.copy()
            if env:
                run_env.update(env)

            result = subprocess.run(
                command,
                capture_output=capture_output,
                text=True,
                env=run_env,
                cwd=cwd,
                timeout=timeout
            )

            self.cli.clear_progress()

            return CommandResult(
                success=result.returncode == 0,
                stdout=result.stdout or "",
                stderr=result.stderr or "",
                return_code=result.returncode
            )

        except subprocess.TimeoutExpired:
            self.cli.clear_progress()
            return CommandResult(
                success=False,
                stdout="",
                stderr="Command timed out",
                return_code=-1
            )
        except FileNotFoundError:
            self.cli.clear_progress()
            return CommandResult(
                success=False,
                stdout="",
                stderr=f"Command not found: {command[0]}",
                return_code=-1
            )
        except Exception as e:
            self.cli.clear_progress()
            return CommandResult(
                success=False,
                stdout="",
                stderr=str(e),
                return_code=-1
            )

    def run_interactive(
        self,
        command: list[str],
        description: Optional[str] = None,
        sudo: Optional[bool] = None,
        env: Optional[dict] = None,
        cwd: Optional[str] = None
    ) -> bool:
        """Run a command with interactive output (shows stdout/stderr in real-time)."""
        use_sudo = sudo if sudo is not None else self.use_sudo

        if use_sudo and os.geteuid() != 0:
            command = ["sudo"] + command

        if description:
            self.cli.print_info(description)

        if self.dry_run:
            self.cli.print_info(f"[DRY RUN] {' '.join(command)}")
            return True

        try:
            run_env = os.environ.copy()
            if env:
                run_env.update(env)

            result = subprocess.run(
                command,
                env=run_env,
                cwd=cwd
            )
            return result.returncode == 0

        except Exception as e:
            self.cli.print_error(f"Erreur: {e}")
            return False

    def check_command_exists(self, command: str) -> bool:
        """Check if a command exists in PATH."""
        return shutil.which(command) is not None

    def get_command_path(self, command: str) -> Optional[str]:
        """Get the full path to a command."""
        return shutil.which(command)

    def get_command_version(self, command: str, version_arg: str = "--version") -> Optional[str]:
        """Get the version of a command."""
        result = self.run([command, version_arg], capture_output=True)
        if result.success:
            # Return first line of output
            output = result.stdout.strip() or result.stderr.strip()
            return output.split("\n")[0] if output else None
        return None

    def ensure_directory(self, path: str, mode: int = 0o755) -> bool:
        """Ensure a directory exists, creating it if necessary."""
        try:
            if self.dry_run:
                self.cli.print_info(f"[DRY RUN] mkdir -p {path}")
                return True

            os.makedirs(path, mode=mode, exist_ok=True)
            return True
        except Exception as e:
            self.cli.print_error(f"Impossible de creer {path}: {e}")
            return False

    def download_file(
        self,
        url: str,
        destination: str,
        description: Optional[str] = None
    ) -> bool:
        """Download a file using curl or wget."""
        if description:
            self.cli.print_progress(description)

        # Try curl first
        if self.check_command_exists("curl"):
            result = self.run(
                ["curl", "-fsSL", "-o", destination, url],
                sudo=False
            )
            if result.success:
                self.cli.clear_progress()
                return True

        # Fall back to wget
        if self.check_command_exists("wget"):
            result = self.run(
                ["wget", "-q", "-O", destination, url],
                sudo=False
            )
            if result.success:
                self.cli.clear_progress()
                return True

        self.cli.clear_progress()
        self.cli.print_error("Ni curl ni wget n'est disponible")
        return False
